src/libcore/intrinsics.rs

```diff
@@ -660,6 +660,18 @@ extern "rust-intrinsic" {
     /// Executes a breakpoint trap, for inspection by a debugger.
     pub fn breakpoint();
 
+    /// Returns `true` during constant evaluation and `false` otherwise.
+    ///
+    /// # Safety
+    ///
+    /// This intrinsic allows breaking referential-transparency in `const fn`
+    /// and is therefore `unsafe`.
+    ///
+    /// Code that uses this intrinsic must be extremely careful to ensure that
+    /// `const fn`s remain referentially-transparent independently of when they
+    /// are evaluated.
+    pub fn is_const_eval() -> bool;
```

Comment:
Loads of platforms will just want to default to the pure rust impl. Maybe that's the naming direction we should go into?

Reply:
What pure rust impl ?

Reply:
Well the function is used to switch between pure rust code and something const eval can't handle. Wasm also can't handle some stuff, so they may often choose to use the pure rust impl, too.

Reply:
That something that const eval can't handle might also be "pure rust code".

Reply:
For packed_simd, many of the codegen workarounds are just calls to any of the thousands of `core::arch` intrinsics.

Reply:
right...

Reply:
If what you are proposing is that it should be possible for a user, e.g., via a compiler flag, to force this intrinsic to always return `true`, then I think that might be doable, and don't see any fundamental problems with allowing that.